File: day3.lua
require('util')

function get_priority(letter)
	code = string.byte(letter)
	offset = 96
	if code > 64 and code < 97 then
		offset = 38
	end
	return code - offset
end

function calc_priority(line)
	l = #line
	for i = 1,l/2 do
		for j = l/2+1,l do
			char1 = string.sub(line,i,i)
			char2 = string.sub(line,j,j)
			if char1 == char2 then
				return get_priority(char1)
			end
		end
	end
end


total = 0
for line in io.lines('day3.txt') do
	total = total + calc_priority(line)
end

-- day 3 part 1
print(total)

function compare(s1,s2)
	common = ''
	for i = 1,#s1 do
		for j = 1,#s2 do
			l1 = string.sub(s1,i,i)
			l2 = string.sub(s2,j,j)
			if l1 == l2 then
				common = common .. l1
			end	
		end
	end
	return common
end


total = 0
group = {}
count = 0
for line in io.lines('day3.txt') do
	count = count + 1
	group[count] = line
	if count == 3 then
		total = total + get_priority(compare(compare(group[1],group[2]),compare(group[2],group[3])):sub(1,1))
		count = 0
		group = {}
	end 
end

print(total)

File: day3.c
#include<stdio.h>
#include<string.h>

int get_priority(char c){return c>96?c-96:c-64+26;}

char buffer[3][200];

int compare(){
        for(int i=0;i<strlen(buffer[0]);i++){
                for(int j=0;j<strlen(buffer[1]);j++){
                        for(int k=0;k<strlen(buffer[2]);k++){
                                if(buffer[0][i]==buffer[1][j]&&buffer[1][j]==buffer[2][k]){
                                        return get_priority(buffer[0][i]);
                                }
                        }
                }
        }
        return 0;
}

int main(){
        int value = 0,count = 0,common = 0;
        FILE* input = fopen("day3.txt","r");
        while(!feof(input)){
                fgets(buffer[count],199,input);
                int len = strlen(buffer[count]);
                for(int i=0;i<len/2;i++){
                        for(int j=len/2;j<len;j++){
                                if(buffer[count][i] == buffer[count][j]){
                                        value += get_priority(buffer[count][i]);
                                        goto next;
                                }
                        }
                }
                next:
                if(count==2){
                        count = 0;
                        common += compare();
                }else{
                        count += 1;
                }
        }
        printf("Part 1: %d\n",value);
        printf("Part 2: %d\n",common);
        fclose(input);
}
